### PR TITLE
Added openstack node template support

### LINKED ISSUE #27: NodeTemplate to support OpenStackConfig
Rancher2 provider started to support NodeTemplates. It would be  nice to support OpenStack as well.

### PR DESCRIPTION
**Pull request by rodcloutier**
This PR intends to fix #27

### CODE REVIEW

**Comment by rodcloutier** [comment]
The build fails but the error seems to be on master. 
`make build-rancher` passes on the branch but fails on master

**Comment by rawmind0** [comment]
Hi @rodcloutier , thanks for the PR.

build is failing to check that code complies with gofmt requirements on file 
`./rancher2/resource_rancher2_node_template.go`. Please run `make fmt` and commit updated `./rancher2/resource_rancher2_node_template.go`

**Comment by rodcloutier** [comment]
I did but it is fine. Run `make fmt` on the master branch and you will get the error...
The error is here https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_node_template.go#L447

My commit actually fixes the issue. There is something weird going on...

**Comment by rawmind0** [comment]
Thanks for the update. I'm going to review the PR taking it into account.

File: rancher2/node_template_openstack_config.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+			Optional: true,
+		},
+		"config_drive": &schema.Schema{
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"domain_id": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"domain_name"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+			Optional:      true,
+		},
+		"domain_name": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"domain_id"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+			Optional:      true,
+		},
+		"endpoint_type": &schema.Schema{
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"flavor_id": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"flavor_name"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+			Optional:      true,
+		},
+		"flavor_name": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"flavor_id"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+			Optional:      true,
+		},
+		"floatingip_pool": &schema.Schema{
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"image_id": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"image_name"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+			Optional:      true,
+		},
+		"image_name": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"image_id"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+			Default:  "22",
+		},
+		"ssh_user": &schema.Schema{
+			Type:     schema.TypeString,
+			Optional: true,
+			Default:  "root",
+		},
+		"tenant_id": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"tenant_name"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

File: rancher2/node_template_openstack_config.go [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+			Optional:      true,
+		},
+		"tenant_name": &schema.Schema{
+			Type:          schema.TypeString,
+			ConflictsWith: []string{"tenant_id"},
```
Comment (rawmind0) [inline rancher2/node_template_openstack_config.go]: Please remove `ConflictsWith` definition..

**Review by rawmind0 — CHANGES_REQUESTED** [review]
`ConflictsWith` is failing on execution of the provider. It seems that `ConflictsWith` is not working fine on nested structures. https://github.com/hashicorp/terraform/issues/11387, https://github.com/hashicorp/terraform/issues/11101

**Comment by rawmind0** [comment]
@rodcloutier , i've tested and reviewed your PR and it's not working fine. Please, take a look to changes requested. 

There are also another changes that needs to be implemented. 
- As `openstack` node driver is not activated by default on Rancher, wait for node driver activation is done is needed. 
- On `resourceRancher2NodeTemplateUpdate` function on `switch driver` openstack case is missed. I also noticed that digitalocean case was wrong and i take profit to address it.

Please, take a look to diff content on `rancher2/resource_rancher2_node_template.go`

```
diff --git a/rancher2/resource_rancher2_node_template.go b/rancher2/resource_rancher2_node_template.go
index fa03d70..37a927b 100644
--- a/rancher2/resource_rancher2_node_template.go
+++ b/rancher2/resource_rancher2_node_template.go
@@ -397,6 +397,19 @@ func resourceRancher2NodeTemplateCreate(d *schema.ResourceData, meta interface{}
                return err
        }

+       stateConf := &resource.StateChangeConf{
+               Pending:    []string{},
+               Target:     []string{"active"},
+               Refresh:    nodeDriverStateRefreshFunc(client, nodeTemplate.Driver),
+               Timeout:    d.Timeout(schema.TimeoutCreate),
+               Delay:      1 * time.Second,
+               MinTimeout: 3 * time.Second,
+       }
+       _, waitErr := stateConf.WaitForState()
+       if waitErr != nil {
+               return fmt.Errorf("[ERROR] waiting for node driver (%s) to be activated: %s", nodeTemplate.Driver, waitErr)
+       }
+
        newNodeTemplate := &NodeTemplate{}

        err = client.APIBaseClient.Create(managementClient.NodeTemplateType, nodeTemplate, newNodeTemplate)
@@ -404,7 +417,7 @@ func resourceRancher2NodeTemplateCreate(d *schema.ResourceData, meta interface{}
                return err
        }

-       stateConf := &resource.StateChangeConf{
+       stateConf = &resource.StateChangeConf{
                Pending:    []string{},
                Target:     []string{"active"},
                Refresh:    nodeTemplateStateRefreshFunc(client, newNodeTemplate.ID),
@@ -412,7 +425,7 @@ func resourceRancher2NodeTemplateCreate(d *schema.ResourceData, meta interface{}
                Delay:      1 * time.Second,
                MinTimeout: 3 * time.Second,
        }
-       _, waitErr := stateConf.WaitForState()
+       _, waitErr = stateConf.WaitForState()
        if waitErr != nil {
                return fmt.Errorf("[ERROR] waiting for node template (%s) to be created: %s", newNodeTemplate.ID, waitErr)
        }
@@ -485,7 +498,9 @@ func resourceRancher2NodeTemplateUpdate(d *schema.ResourceData, meta interface{}
        case azureConfigDriver:
                update["azureConfig"] = expandAzureConfig(d.Get("azure_config").([]interface{}))
        case digitaloceanConfigDriver:
-               update["digitaloceanConfig"] = expandAzureConfig(d.Get("digitalocean_config").([]interface{}))
+               update["digitaloceanConfig"] = expandDigitaloceanConfig(d.Get("digitalocean_config").([]interface{}))
+       case openstackConfigDriver:
+               update["openstackConfig"] = expandOpenstackConfig(d.Get("openstack_config").([]interface{}))
        }

        newNodeTemplate := &NodeTemplate{}
```

- Please add `openstack` node driver tests. Take a look to diff content on `rancher2/resource_rancher2_node_template_test.go`

```
diff --git a/rancher2/resource_rancher2_node_template_test.go b/rancher2/resource_rancher2_node_template_test.go
index 4628170..a3b6648 100644
--- a/rancher2/resource_rancher2_node_template_test.go
+++ b/rancher2/resource_rancher2_node_template_test.go
@@ -120,6 +120,52 @@ resource "rancher2_node_template" "foo" {
        access_token = "XXXXXXXXXXXXXXXXXXXX"
   }
 }
+`
+
+       testAccRancher2NodeTemplateConfigOpenstack = `
+resource "rancher2_node_template" "foo" {
+  name = "foo"
+  description = "Terraform node driver openstack acceptance test"
+  openstack_config {
+       username = "XXXXXXXXXXXXXXXXXXXX"
+       password = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+       image_name =  "image-XXXXXXXX"
+       region = "XX-west-1"
+       flavor_name = "flavor-XXXXXXXX"
+       auth_url = "http://XXXXXXXX"
+       availability_zone = "zone-XXXXXXXX"
+  }
+}
+`
+       testAccRancher2NodeTemplateUpdateConfigOpenstack = `
+resource "rancher2_node_template" "foo" {
+  name = "foo2"
+  description = "Terraform node driver openstack acceptance test - updated"
+  openstack_config {
+       username = "XXXXXXXXXXXXXXXXXXXX"
+       password = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+       image_name =  "image-YYYYYYYY"
+       region = "XX-west-1"
+       flavor_name = "flavor-XXXXXXXX"
+       auth_url = "http://XXXXXXXX"
+       availability_zone = "zone-XXXXXXXX"
+  }
+}
+ `
+       testAccRancher2NodeTemplateRecreateConfigOpenstack = `
+resource "rancher2_node_template" "foo" {
+  name = "foo"
+  description = "Terraform node driver openstack acceptance test"
+  openstack_config {
+       username = "XXXXXXXXXXXXXXXXXXXX"
+       password = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+       image_name =  "image-XXXXXXXX"
+       region = "XX-west-1"
+       flavor_name = "flavor-XXXXXXXX"
+       auth_url = "http://XXXXXXXX"
+       availability_zone = "zone-XXXXXXXX"
+  }
+}
 `
 )

@@ -312,6 +358,69 @@ func TestAccRancher2NodeTemplate_disappears_Digitalocean(t *testing.T) {
        })
 }

+func TestAccRancher2NodeTemplate_basic_Openstack(t *testing.T) {
+       var nodeTemplate *NodeTemplate
+
+       name := testAccRancher2NodeTemplateType + ".foo"
+       resource.Test(t, resource.TestCase{
+               PreCheck:     func() { testAccPreCheck(t) },
+               Providers:    testAccProviders,
+               CheckDestroy: testAccCheckRancher2NodeTemplateDestroy,
+               Steps: []resource.TestStep{
+                       resource.TestStep{
+                               Config: testAccRancher2NodeTemplateConfigOpenstack,
+                               Check: resource.ComposeTestCheckFunc(
+                                       testAccCheckRancher2NodeTemplateExists(name, nodeTemplate),
+                                       resource.TestCheckResourceAttr(name, "name", "foo"),
+                                       resource.TestCheckResourceAttr(name, "description", "Terraform node driver openstack acceptance test"),
+                                       resource.TestCheckResourceAttr(name, "driver", openstackConfigDriver),
+                                       resource.TestCheckResourceAttr(name, "openstack_config.0.image_name", "image-XXXXXXXX"),
+                               ),
+                       },
+                       resource.TestStep{
+                               Config: testAccRancher2NodeTemplateUpdateConfigOpenstack,
+                               Check: resource.ComposeTestCheckFunc(
+                                       testAccCheckRancher2NodeTemplateExists(name, nodeTemplate),
+                                       resource.TestCheckResourceAttr(name, "name", "foo2"),
+                                       resource.TestCheckResourceAttr(name, "description", "Terraform node driver openstack acceptance test - updated"),
+                                       resource.TestCheckResourceAttr(name, "driver", openstackConfigDriver),
+                                       resource.TestCheckResourceAttr(name, "openstack_config.0.image_name", "image-YYYYYYYY"),
+                               ),
+                       },
+                       resource.TestStep{
+                               Config: testAccRancher2NodeTemplateRecreateConfigOpenstack,
+                               Check: resource.ComposeTestCheckFunc(
+                                       testAccCheckRancher2NodeTemplateExists(name, nodeTemplate),
+                                       resource.TestCheckResourceAttr(name, "name", "foo"),
+                                       resource.TestCheckResourceAttr(name, "description", "Terraform node driver openstack acceptance test"),
+                                       resource.TestCheckResourceAttr(name, "driver", openstackConfigDriver),
+                                       resource.TestCheckResourceAttr(name, "openstack_config.0.image_name", "image-XXXXXXXX"),
+                               ),
+                       },
+               },
+       })
+}
+
+func TestAccRancher2NodeTemplate_disappears_Openstack(t *testing.T) {
+       var nodeTemplate *NodeTemplate
+
+       resource.Test(t, resource.TestCase{
+               PreCheck:     func() { testAccPreCheck(t) },
+               Providers:    testAccProviders,
+               CheckDestroy: testAccCheckRancher2NodeTemplateDestroy,
+               Steps: []resource.TestStep{
+                       resource.TestStep{
+                               Config: testAccRancher2NodeTemplateConfigDigitalocean,
+                               Check: resource.ComposeTestCheckFunc(
+                                       testAccCheckRancher2NodeTemplateExists(testAccRancher2NodeTemplateType+".foo", nodeTemplate),
+                                       testAccRancher2NodeTemplateDisappears(nodeTemplate),
+                               ),
+                               ExpectNonEmptyPlan: true,
+                       },
+               },
+       })
+}
+
 func testAccRancher2NodeTemplateDisappears(nodeTemplate *NodeTemplate) resource.TestCheckFunc {
        return func(s *terraform.State) error {
                for _, rs := range s.RootModule().Resources {
```

**Comment by rawmind0** [comment]
> I did but it is fine. Run `make fmt` on the master branch and you will get the error...
> The error is here https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_node_template.go#L447
> 
> My commit actually fixes the issue. There is something weird going on...

Build checks are failing due to the golang version. Probably you have installed golang 1.11 or 1.12, but drone is using golang 1.9 for building (as detailed on README.md file).

It seems that go fmt rules have changed from golang 1.9 to 1.11 or 1.12. https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_node_template.go#L447 line is well formatted on 1.9 but wrong on 1.11 or 1.12. 

To fix it, you have 2 options: 
- remove your modification on https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_node_template.go#L447 and let the line as it was.
- Wait for merge PR #72 and rebase your PR.

File: rancher2/resource_rancher2_node_template.go [hunk 9]
```diff
@@ -444,7 +463,7 @@ func resourceRancher2NodeTemplateUpdate(d *schema.ResourceData, meta interface{}
 	}
 
 	update := map[string]interface{}{
-		"name": d.Get("name").(string),
+		"name":                     d.Get("name").(string),
```
Comment (rawmind0) [inline rancher2/resource_rancher2_node_template.go]: To fix build check error, please let line 447 like it was.

```
-		"name":                     d.Get("name").(string),
+		"name": d.Get("name").(string),
```

**Review by rawmind0 — CHANGES_REQUESTED** [review]

**Comment by rodcloutier** [comment]
I guess I will close it in favor of #91.

I know it took a long time before I could revisit the PR but I would of liked to know that this PR was to be superseded by another work in progress...

**Comment by rawmind0** [comment]
@rodcloutier, it wasn't a WIP, but i decided to adapt and include your PR into #91 due to long time of inactivity after review.  Thanks for your PR and sorry about that. 

Superseded by #91

**Pull request closed without merging**